mirdata/msd.py

```diff
@@ -0,0 +1,373 @@
+        self.artist_id = _string(h5song["artist_id"])
+        self.artist_latitude = h5song["artist_latitude"]
+        self.artist_location = _string(h5song["artist_location"])
+        self.artist_longitude = h5song["artist_longitude"]
+        self.artist_mbid = _string(h5song["artist_mbid"])
+        self.artist_name = _string(h5song["artist_name"])
+        self.artist_playmeid = h5song["artist_playmeid"]
+        self.genre = _string(h5song["genre"])
+        self.idx_artist_terms = h5song["idx_artist_terms"]
+        self.idx_similar_artists = h5song["idx_similar_artists"]
+        self.release = _string(h5song["release"])
+        self.release_7digitalid = h5song["release_7digitalid"]
+        self.song_hotttnesss = h5song["song_hotttnesss"]
+        self.song_id = _string(h5song["song_id"])
+        self.title = _string(h5song["title"])
+        self.track_7digitalid = h5song["track_7digitalid"]
+
+    def __repr__(self):
+        return (
+            "MSDMetadataSong("
+            + "artist_id={artist_id}, "
+            + "artist_name={artist_name}, "
+            + "song_id={song_id}, "
+            + "title={title}, [...])"
+        ).format(
+            artist_id=self.artist_id,
+            artist_name=self.artist_name,
+            song_id=self.song_id,
+            title=self.title,
+        )
+
+
+class MSDMusicBrainz(object):
+    def __init__(self, h5mb):
+        self.artist_mbtags = _string_list(h5mb["artist_mbtags"])
+        self.artist_mbtags_count = list(h5mb["artist_mbtags_count"])
+        self.songs = [MSDMusicBrainzSong(s) for s in h5mb["songs"]]
+
+    def __repr__(self):
+        return (
+            "MSDMusicBrainz("
+            + "artist_mbtags={artist_mbtags}, "
+            + "artist_mbtags_count={artist_mbtags_count}, "
+            + "songs={songs})"
+        ).format(
+            artist_mbtags=self.artist_mbtags,
+            artist_mbtags_count=self.artist_mbtags_count,
+            songs=self.songs,
+        )
+
+
+class MSDMusicBrainzSong(object):
+    def __init__(self, h5song):
+        self.idx_artist_mbtags = h5song["idx_artist_mbtags"]
+        self.year = h5song["year"]
+
+    def __repr__(self):
+        return "MSDMusicBrainzSong(year={year})".format(year=self.year)
+
+
+def _string_list(lst):
+    return [_string(x) for x in lst]
+
+
+def _string(s):
+    return s.decode("utf-8")
+
+
+def _filename(track_id, data_home=None):
+    if data_home is None:
+        data_home = utils.get_default_dataset_path(DATASET_DIR)
+
+    # format: 'data/J/D/V/TRJDVIB12903CF9F35.h5'
+    return os.path.join(
+        data_home, "data", track_id[2], track_id[3], track_id[4], track_id + ".h5"
+    )
+
+
+def track_ids(data_home=None):
+    """Generator that yields all MSD track IDs
+
+    Returns:
+        (generator) Track IDs
+    """
+
+    if data_home is None:
+        data_home = utils.get_default_dataset_path(DATASET_DIR)
+
+    for _, _, files in os.walk(data_home):
```

Comment:
I don't like that the output of this function changes depending on what data you have locally.
In the other loaders we're getting the track_ids from the index file. Did you end up including only a subset of the data in the index file? If so, could we store a list of the trackids somewhere, so the user can know what they're missing?

Reply:
I know what you mean, but I think there needs to be a distinction between "small" datasets where the IDs reasonably fit in the repository, and "large" ones like the MSD. The MSD index file, even if we only included filenames, would be ~35MB. If we included hashes as well it'd be well over 100MB, forcing users of mirdata to install git-lfs. I opted for a filesystem-based `track_ids()` function, validated by a probabilistic `validate()` function. It's just a suggestion though, so if you have other ideas I'd happily rewrite this!

Reply:
I agree on the validation side! I just think intuitively a function that gives the dataset's track ids should be deterministic, if possible. Some ideas:
* What about a file that just stores the track ids (no filenames or checksums)?
* I opened an idea for how we deal with the growing size of the `indexes/` folder in #153 - we could do something similar here (download the trackids file from zenodo as part of the download).

Reply:
I like that idea! In that case we could even store the full index on Zenodo. If you have bandwidth to download ~500GB of MSD, you can probably download another ~100MB of index. The `validate` method could get an optional `subsample_percent=1.0` (or something) argument, in case you want faster validation.